src/shaka_scheme/system/base/Data.cpp

```diff
@@ -2,11 +2,12 @@
 // Created by aytas on 8/6/2017.
 //
 
-#include "shaka_scheme/system/base/Data.hpp"
+#include <shaka_scheme/system/exceptions/MissingImplementationException.hpp>
 
+#include "shaka_scheme/system/base/Data.hpp"
 
-shaka::Data::Data(const shaka::Data &other) : type_tag
-                                                  (shaka::Data::Type::INVALID) {
+shaka::Data::Data(const shaka::Data& other) :

```

Comment:
"**Please also review whether my implementation of deep copying of Data to other Data in the copy constructor in Data(const Data& other) is correct. I'm not sure if it is.**"

Assuming that the copy constructor for DataPair does the deep copy correctly for you in the case where Data holds a DataPair (which having looked at it's implementation in DataPair.cpp, it appears to do properly), this copy constructor should behave as expected in all cases. You are constructing the `data_pair` field by calling the copy constructor in `new(&data_pair) shaka::DataPair(other.data_pair);` which appears to obey the correct deep copy semantics as far as I can tell. Perhaps the only other class I would be concerned about is `shaka::String` since you also delegate to its copy constructor in the case that Data holds a string. Assuming that all of the copy constructors that you delegate to in this constructor are coded correctly, this method should perform the appropriate deep copy when needed (when data doesn't just hold some simple, atomic datatype)

Just to confirm, is Data still going to be able to hold Environments as a type of scheme datum? I only ask because the enum tag for `ENVIRONMENT` is still present in data.hpp. In terms of the way that we are designing things this time around, I might conjecture that it is semantically incorrect to have environments be a kind of scheme datum in our system, as there is no construct built into the actual scheme language that allows us to explicitly pass around references to an environment object. The environment is a meta-level construct of the language. Otherwise, everything that I have seen in these change logs looks rock-solid.

Reply:
Not sure about the environment part. But, otherwise, looks good.